### ISSUE TITLE
Vectorize PLCI source-anchored prediction to remove MOSEI utterance-loop bottleneck

### ISSUE THREAD

**Issue by YB123-DT** [YB123-DT/git_gcnet#1]
## Problem

The current PLCI-JEPA predictor performs source/target prediction one valid utterance at a time in Python.

Affected implementation:

- `gcnet_plci_jepa/modules.py`
  - `SourceAnchoredPredictor.forward()`
  - nested `for time in range(length)` / `for item in range(batch)` loops
  - per-utterance `availability[time, item].tolist()`
  - repeated scalar embedding construction and small MLP/kernel launches
- `gcnet_plci_jepa/loss.py`
  - `plci_jepa_loss()` iterates over `List[PLCITargetPrediction]` one target record at a time

Each loop item represents one utterance/segment node, **not one speaker and not one whole dialogue**. Dialogue context has already been encoded by the pre-graph recurrent encoder and Temporal/Speaker GCNet branches before the predictor receives `graph_hidden[L,B,H]`.

## Observed impact

Cross-dataset PLCI sweep configuration:

- CMU-MOSEI
- 100 epochs
- 7 processes on one V100-class 32 GB GPU
- Natural GCNet forward + pattern-balanced auxiliary GCNet forward
- official validation/test calls

Observed after approximately 2 h 8 min:

- only 18–19/100 epochs completed
- approximately 6.7 min/epoch
- approximately 11 h for a group of 7 experiments on one GPU
- no OOM, traceback, or dead process
- GPU utilization remains high, but work is fragmented into many small operations

MOSEI contains roughly 22.8k utterance/segment nodes, so the per-node Python and synchronization cost is much more visible than on MOSI or IEMOCAP.

## Root cause

For every valid `[time, batch]` position, the current predictor:

1. copies a three-element availability vector to Python through `.tolist()`;
2. identifies one of six patterns;
3. creates scalar embedding-index tensors;
4. calls canonicalizer/base/context/innovation MLPs on individual vectors;
5. appends a Python dataclass record;
6. later loops over all records again in the loss.

This prevents the GPU from processing pattern-homogeneous utterances as large matrices and may introduce repeated GPU/CPU synchronization.

## Proposed implementation

Add a vectorized backend while preserving the existing implementation as a reference oracle.

### 1. Flatten valid utterances once

```python
valid = umask.transpose(0, 1).bool()
flat_hidden = graph_hidden[valid]       # [N, H]
flat_pattern = availability[valid]      # [N, 3]
flat_latents = {
    name: latent[valid]                 # [N, d]
    for name, latent in student_latents.items()
}
```

### 2. Compute pattern IDs entirely on device

```python
binary_code = (
    4 * flat_pattern[:, 0].long()
    + 2 * flat_pattern[:, 1].long()
    + flat_pattern[:, 2].long()
)
lookup = torch.tensor(
    [-1, 2, 1, 5, 0, 4, 3, -1],
    device=flat_pattern.device,
)
pattern_id = lookup[binary_code]
```

Mapping:

- `100 -> A`
- `010 -> T`
- `001 -> V`
- `110 -> AT`
- `101 -> AV`
- `011 -> TV`

### 3. Batch by six patterns

Loop over at most six pattern groups, not over every utterance. For a dual-source group such as `AT -> V`, compute in batch:

- `g_A`, `g_T`
- shared current-view context `c_(AT->V)(h_J)`
- base anchors `beta_(A->V)`, `beta_(T->V)`
- ordered innovations `Delta_(T|A->V)`, `Delta_(A|T->V)`
- the two normalized paths

All tensors should have leading dimension `N_pattern`.

### 4. Grouped prediction container and vectorized loss

Replace the per-utterance list in the optimized backend with grouped records:

```python
@dataclass
class PLCIGroupPrediction:
    utterance_indices: Tensor   # [Ng]
    target_modality: int
    paths: Tensor               # [Ng, num_paths, d]
```

Gather teacher targets once per group and compute cosine loss in matrix form. Use `index_add_`/scatter operations to preserve the existing hierarchy:

1. average paths within a target;
2. average targets within an utterance;
3. average valid utterances.

## Non-goals

This optimization must not change:

- GCNet topology or dialogue boundaries;
- speaker handling or `qmask`;
- six-pattern sampling;
- source-only base equations;
- shared context equations;
- ordered innovation equations;
- normalization;
- teacher targets or EMA update;
- loss weighting;
- inference computation;
- parameter names, shapes, count, or initialization.

## Required equivalence tests

Keep both backends during validation:

```text
reference  = current utterance-loop implementation
vectorized = proposed grouped implementation
```

Using identical parameters and inputs, test:

1. all six patterns in one batch;
2. singleton and dual-source targets;
3. mixed dialogue lengths and padding;
4. IEMOCAP two-speaker inputs;
5. MOSI/MOSEI one-speaker inputs;
6. CPU FP32;
7. CUDA FP32.

Forward tolerance:

```text
max absolute prediction error < 1e-6
absolute PLCI loss error < 1e-6
```

Backward tolerance:

```text
max gradient error < 1e-5
```

Compare gradients for:

- student projectors;
- source canonicalizers;
- base predictor;
- graph-context correction;
- conditional innovation;
- GCNet hidden input.

Also verify exact equality of:

- parameter keys;
- parameter shapes and count;
- EMA update result;
- pattern/target/path counts.

## Benchmark requirement

After equivalence passes, benchmark reference versus vectorized implementations on the same real MOSEI batch:

- predictor forward time;
- PLCI loss time;
- total Natural + Auxiliary forward/backward time;
- peak GPU memory;
- utterances/second.

Report warm-up and at least 20 timed iterations with CUDA synchronization around measurements.

## Experiment compatibility warning

Do **not** mix the vectorized backend into an already running formal sweep. Even if mathematical equivalence passes, a formal result table must use one implementation revision throughout. Either:

- finish the current sweep with the reference backend, then use vectorization for later experiments; or
- restart the entire affected sweep under the vectorized revision.
